### ISSUE TITLE
"Each GPU has a different forward pass (in terms of used parameters)" slows down training

### ISSUE THREAD

**Issue by formiel** [NVIDIA/apex#676]
Hello,

I have a model where in each forward pass, layers are randomly skipped with some probability, something like this:

```
def forward(self, x):
    p = 0.2
    for i in range(self.n_layers):
        # drop this layer with probability p
        random_number = random.uniform(0, 1)
        if self.training and (random_number < p):
            continue
        x = self.layers[i](x)
    return x
```
Theoretically, higher p should result in faster training as the number of skipped layers is higher. However, when training this model using `apex.parallel.DistributedDataParallel`, I observed a very significant slowdown (when p > 0). I try timing the forward passes and indeed, they were faster when p is higher, so I think the issue lies in the backward or the gradient gathering steps.

Note that this issue does not occur when using `torch.nn.parallel.DistributedDataParallel` (I had to set `find_unused_parameters = True` for it to work): faster training for higher p.

Could you please help checking this?
When all layers are used (i.e. p = 0), `apex` was faster than `torch.nn` (especially on multiple nodes), so I hope that I can use `apex` for all my experiments.
Thank you very much in advance!

**Comment by formiel** [comment]
Sorry I would like to clarify that the slowdown occurred for multi-node training only. Below is a small benchmark:

**Single node with 2 GPUs:**
- p = 0.0: 52.69 sentences/s -  4057.50 words/s
- p = 0.2: 58.20 sentences/s -  4394.05 words/s

**Two nodes with 2 GPUs each:**
- p = 0.0: 36.51 sentences/s -  2770.76 words/s
- p = 0.2: 14.30 sentences/s -  1098.06 words/s

Thanks.

**Comment by BramVanroy** [comment]
It has been mentioned a couple of times in the issues here on apex that in the current state torch DDP is preferred over apex's implementation anyway, so it might be best to stick with that. When apex goes upstream, I don't think the apex DDP will be ported either, and torch's DDP will stay the default anyway.

If you really want to dig into why this is happening, I can't help with that.
